modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/CreateIndexOnInvalidDataTypeTest.java

```diff
@@ -0,0 +1,198 @@
+     * - try to create index for the new field - exception must be throw;
+     * - checks that index isn't created.
+     */
+    @Test
+    public void testCreateIndexOnInvalidData() throws Exception {
+        startGrid();
+
+        grid().cluster().state(ClusterState.ACTIVE);
+
+        IgniteCache<Integer, Value> c = grid().createCache(new CacheConfiguration<Integer, Value>()
+            .setName("test")
+            .setSqlSchema("PUBLIC")
+            .setQueryEntities(Collections.singleton(new QueryEntity(Integer.class, Value.class)
+                    .setTableName("TEST")
+                )
+            )
+            .setBackups(1)
+            .setAffinity(new RendezvousAffinityFunction(false, 10)));
+
+        for (int i = 0; i < KEY_CNT; ++i)
+            c.put(i, new Value(i));
+
+        sql("ALTER TABLE TEST ADD COLUMN (VAL_DATE DATE)");
+
+        sql("CREATE INDEX TEST_VAL_INT_IDX ON TEST(VAL_INT)");
+
+        GridTestUtils.assertThrowsAnyCause(log, () -> {
+                sql("CREATE INDEX TEST_VAL_DATE_IDX ON TEST(VAL_DATE)");
+
+                return null;
+            },
+            IgniteSQLException.class, "java.util.Date cannot be cast to java.sql.Date");
+
+        // Wait for node stop if it is initiated by FailureHandler
+        U.sleep(1000);
+
+        List<List<?>> res = sql("SELECT val_int FROM TEST where val_int > -1").getAll();
+
+        assertEquals(KEY_CNT, res.size());
+
+        GridTestUtils.assertThrowsAnyCause(log, () -> {
+                sql("DROP INDEX TEST_VAL_DATE_IDX");
+
+                return null;
+            },
+            IgniteSQLException.class, "Index doesn't exist: TEST_VAL_DATE_IDX");
+    }
+
+    /**
+     * Check case when row with invalid field is added.
+     * Test steps:
+     * - create table;
+     * - create two index;
+     * - try add entry - exception must be thrown;
+     * - remove the index for field with invalid type;
+     * - check that select query that uses the index for valid field is successful.
+     */
+    @Test
+    public void testAddInvalidDataToIndex() throws Exception {
+        startGrid();
+
+        grid().cluster().state(ClusterState.ACTIVE);
+
+        sql("CREATE TABLE TEST (ID INT PRIMARY KEY, VAL_INT INT, VAL_DATE DATE) " +
+            "WITH \"CACHE_NAME=test,VALUE_TYPE=ValueType0\"");
+
+        sql("CREATE INDEX TEST_0_VAL_DATE_IDX ON TEST(VAL_DATE)");
+        sql("CREATE INDEX TEST_1_VAL_INT_IDX ON TEST(VAL_INT)");
+
+        BinaryObjectBuilder bob = grid().binary().builder("ValueType0");
+
+        bob.setField("VAL_INT", 10);
+        bob.setField("VAL_DATE", new java.util.Date());
+
+        GridTestUtils.assertThrowsAnyCause(log, () -> {
+                grid().cache("test").put(0, bob.build());
+
+                return null;
+            },
+            CachePartialUpdateCheckedException.class, "Failed to update keys");
+
+        sql("DROP INDEX TEST_0_VAL_DATE_IDX");
```

Comment:
let's verify that after index was dropped you could insert values without any error